qemu/tests/virtio_console.py

```diff
@@ -37,6 +39,40 @@ def __set_exit_event():
     EXIT_EVENT.set()
 
 
+def add_unused_chardev_cmdline(vm, params):
+    """
+    Add '-chardev xxx' cmdline to vm without serial device utilize it
+
+    :param vm: VM object to be operated
+    :param params: Dictionary with the test parameters
+    :return list of added CharDevice object
+    """
+    qemu_binary = utils_misc.get_qemu_binary(params)
+    devices = qcontainer.DevContainer(qemu_binary, vm.name,
+                                      params.get('strict_mode'),
+                                      params.get(
+                                          'workaround_qemu_qmp_crash'),
+                                      params.get('allow_hotplugged_vm'))
+    char_devices = params['extra_chardevs'].split()
+    host = params.get('chardev_host', '127.0.0.1')
+    free_ports = utils_misc.find_free_ports(
+        5000, 6000, len(char_devices), host)
+    devices = []
+    for index, chardev in enumerate(char_devices):
+        chardev_param = params.object_params(chardev)
+        file_name = vm.get_serial_console_filename(chardev)
+        backend = chardev_param.get('chardev_backend',
+                                    'unix_socket')
+        if backend in ['udp', 'tcp_socket']:
+            chardev_param['chardev_host'] = host
+            chardev_param['chardev_port'] = free_ports[index]
+        device = devices.chardev_define_by_params(chardev, chardev_param, file_name)
```

Comment:
Test case:
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci: ERROR: 'list' object has no attribute 'chardev_define_by_params' (3.46 s)

02:05:23 ERROR| File "/home/kar/kar/workspace/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_serial_hotplug_port_pci.py", line 18, in run
02:05:23 ERROR| char_devices = add_unused_chardev_cmdline(vm, params)
02:05:23 ERROR| File "/home/kar/kar/workspace/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/virtio_console.py", line 69, in add_unused_chardev_cmdline
02:05:23 ERROR| device = devices.chardev_define_by_params(chardev, chardev_param, file_name)
02:05:23 ERROR| AttributeError: 'list' object has no attribute 'chardev_define_by_params

Reply:
@nanliu-r Updated, thanks.